File: haskell/2021/day_03.hs
import qualified Data.Char as Char

-- | Convert a list of bits into an integer.
binToInt :: [Int] -> Int
binToInt = foldl1 (\acc b -> 2 * acc + b)

-- | Map the inner elements from (0,1) to (-1,1),
-- then fold the lists via element-wise summation.
getDeltas :: [[Int]] -> [Int]
getDeltas = foldl1 (zipWith (+)) . (map . map) (\b -> b * 2 - 1)

-- | Compute the power consumption of the submarine.
findPower :: [[Int]] -> Int
findPower bins = gamma * epsilon
    where
    deltas = getDeltas bins
    gamma = binToInt [if d > 0 then 1 else 0 | d <- deltas]
    epsilon = binToInt [if d < 0 then 1 else 0 | d <- deltas]

-- | Compute the o2 (True) or co2 (False) rating from a list of binary strings.
findRating :: Bool -> Int -> [[Int]] -> Int
findRating _ _ [] = error "List is empty"
findRating _ _ [bin] = binToInt bin
findRating o2 i bins = findRating o2 (i + 1) $ filter predicate bins
    where
    deltas = getDeltas bins
    keep = if (deltas!!i >= 0) == o2 then 1 else 0
    predicate = \bin -> bin!!i == keep

main :: IO ()
main = do
    text <- readFile "../../input/2021/input_03.txt"
    let bins = (map . map) Char.digitToInt $ lines text
    putStrLn $ "Part 1: " ++ show (findPower bins)
    let oxygen = findRating True 0 bins
    let co2 = findRating False 0 bins
    putStrLn $ "Part 2: " ++ show (oxygen * co2)

File: rust/2021/src/bin/day_03.rs
use std::fs;

fn get_deltas(lines: &[&str]) -> Vec<i32> {
    let mut deltas: Vec<i32> = vec![0; lines[0].len()];
    for line in lines {
        for (i, c) in line.bytes().enumerate() {
            deltas[i] += if c == b'1' { 1 } else { -1 }
        }
    }
    deltas
}

fn find_rating(lines: &[&str], o2: bool) -> u32 {
    let mut lines = Vec::from(lines);
    let mut i = 0;
    while lines.len() > 1 {
        let deltas = get_deltas(&lines);
        // If 1 is most frequent or equal, keep 1 for o2 and 0 for co2
        let keep = if (deltas[i] >= 0) == o2 { b'1' } else { b'0' };
        lines = lines.into_iter().filter(|s| s.as_bytes()[i] == keep).collect();
        i += 1;
    }
    u32::from_str_radix(lines[0], 2).unwrap()
}

fn main() {
    let input = fs::read_to_string("../input/2021/input_03.txt").unwrap();
    let lines: Vec<_> = input.lines().collect();
    let tracker = get_deltas(&lines);
    let gamma: u32 = tracker.iter().rev().enumerate()
        .map(|(i, &b)| if b > 0 { 1 << i } else { 0 })
        .sum();
    let epsilon = !gamma & (1 << tracker.len()) - 1;
    println!("Part 1: {}", gamma * epsilon);
    let oxygen = find_rating(&lines, true);
    let co2 = find_rating(&lines, false);
    println!("Part 2: {}", oxygen * co2);
}
